src/SQLProvider/Providers.Firebird.fs

```diff
@@ -0,0 +1,962 @@
+﻿namespace FSharp.Data.Sql.Providers
+
+open System
+open System.Collections.Concurrent
+open System.Collections.Generic
+open System.Data
+open FSharp.Data.Sql
+open FSharp.Data.Sql.Schema
+open FSharp.Data.Sql.Common
+open FSharp.Data.Sql.Transactions
+
+module Firebird =
+    let mutable resolutionPath = String.Empty
+    let mutable owner = String.Empty
+    let mutable referencedAssemblies = [||]
+
+    let assemblyNames = [
+        "FirebirdSql.Data.FirebirdClient.dll"
+    ]
+
+    let assembly =
+        lazy Reflection.tryLoadAssemblyFrom resolutionPath referencedAssemblies assemblyNames
+
+    let findType name =
+        match assembly.Value with
+        | Choice1Of2(assembly) -> assembly.GetTypes() |> Array.find(fun t -> t.Name = name)
+        | Choice2Of2(paths, errors) ->
+           let details = 
+                match errors with 
+                | [] -> "" 
+                | x -> Environment.NewLine + "Details: " + Environment.NewLine + String.Join(Environment.NewLine, x)
+           failwithf "Unable to resolve assemblies. One of %s (e.g. from Nuget package Firebird.Data) must exist in the paths: %s %s %s"
+                (String.Join(", ", assemblyNames |> List.toArray))
+                Environment.NewLine
+                (String.Join(Environment.NewLine, paths |> Seq.filter(fun p -> not(String.IsNullOrEmpty p))))
+                details
+
+    let connectionType =  lazy (findType "FbConnection")
+    let commandType =     lazy (findType "FbCommand")
+    let parameterType =   lazy (findType "FbParameter")
+    let enumType =        lazy (findType "FbDbType")
+    let getSchemaMethod = lazy (connectionType.Value.GetMethod("GetSchema",[|typeof<string>; typeof<string[]>|]))
+    let paramEnumCtor   = lazy parameterType.Value.GetConstructor([|typeof<string>;enumType.Value|])
+    let paramObjectCtor = lazy parameterType.Value.GetConstructor([|typeof<string>;typeof<obj>|])
+
+    let getSchema name (args:string[]) (conn:IDbConnection) =
+        getSchemaMethod.Value.Invoke(conn,[|name; args|]) :?> DataTable
+
+    let mutable typeMappings = []
+    let mutable findClrType : (string -> TypeMapping option)  = fun _ -> failwith "!"
+    let mutable findDbType : (string -> TypeMapping option)  = fun _ -> failwith "!"
+
+    let rec fieldNotation (al:alias) (c:SqlColumnType) =
+        let colSprint = 
+            match String.IsNullOrEmpty(al) with
+            | true -> sprintf "%s"
+            | false -> sprintf "%s.%s" al
+        match c with
+        // Custom database spesific overrides for canonical functions:
+        | SqlColumnType.CanonicalOperation(cf,col) ->
+            let column = fieldNotation al col
+            match cf with
+            // String functions
+            | Replace(SqlStr(searchItm),SqlStrCol(al2, col2)) -> sprintf "REPLACE(%s,'%s',%s)" column searchItm (fieldNotation al2 col2)
```

Comment:
These commands should correspond the database specific commands, in this case listed here: https://firebirdsql.org/refdocs/langrefupd21-intfunc.html